### DISCUSSION TITLE
Support oAuth flow in JS SDK

### DISCUSSION THREAD

**Discussion by xesjkeee** [directus/directus#4463] [Feature Requests]
Hi!

I've backend-app with directus and frontend-app where I rendered an articles via REST API,  
also I've oauth to log in into admin app. Email & Password are not used.

So how can I authenticate users on frontend-app with sdk-js if I using oauth provider (SSO)?

I found login method, but it requires email and password, no chance 😞 

https://github.com/directus/directus/blob/main/packages/sdk-js/src/handlers/auth.ts#L63

**Comment by miccoh1994** [comment]
Hey @xesjkeee I did some experimenting on this myself yesterday using  create-react-app, directus, and Auth0 (auth0-react lib). You can basically use the Proxy to MITM yourself and retrieve the AuthCode to submit to, /auth/oauth/:provider/callback?code=interceptedcode in Directus to get a directus Access and Refresh Token. How you store this token is up to you. Also I would be wary to do this, I haven't tried with PKCE enabled yet. Which I assume would be critical to ensure the security of the token retrieval since the Authorization Code is passed in the url query :)

**Reply by miccoh1994** [reply]
I used nginx to proxy requests *

**Comment by wc-matteo** [comment]
can this be changed to a feature request? And make clear in the title that it's about using oauth with the js sdk? @rijkvanzanten

**Reply by NilsBaumgartner1994** [reply]
That would be something i would also need. Or as alternative, that the redirect can add the refresh token in the url instead of cookie

**Comment by nodeworks** [comment]
As a follow-up, here's how I was able to implement this:

1. Create a user that you use for creating / logging in users via GraphQL, I created a "public" user with the admin role.
2. Handle OAuth logic from the front-end (I'm using NextAuth with Next JS).
3. Once you get the token from the front-end OAuth, make a login query to the directus API using the public user you created in step 1, this will response with a access token you can use to subsequently hit another custom directus endpoint which I provide below.

I created a custom endpoint at `extensions/endpoints/oauth2/index.js`

This is the endpoint I hit from the front-end to create a session / create a user and get the token and send it back to the front-end. If you use this you might want to change the role, since I hard-coded a role in there.

```javascript
module.exports = function registerEndpoint(router, { services, exceptions, database, getSchema }) {
  router.post('/azure', async (req, res, next) => {
    const { ForbiddenException } = exceptions;
    const { AuthenticationService, UsersService } = services;

    if (!req?.token) {
      return next(new ForbiddenException());
    }

    const schema = await getSchema();
    const email = req?.body?.email.toLowerCase()
    const user = await database
      .select('id', 'status')
      .from('directus_users')
      .whereRaw('LOWER(??) = ?', ['email', email])
      .first();

    const accountability = {
      ip: req.ip,
      userAgent: req.get('user-agent'),
      role: null,
    };

    const authenticationService = new AuthenticationService({
      accountability: accountability,
      schema: req.schema
    });

    if (!user) {
      const usersService = new UsersService({
        schema,
        knex: database
      });

      await usersService.createOne({
        email,
        role: '64CFF9A9-9A5F-48F1-80D1-FF44F4CED284',
        first_name: req?.body?.firstName,
        last_name: req?.body?.lastName
      })

      const { accessToken, refreshToken, expires } = await authenticationService.authenticate({
        email
      });

      res.send({
        accessToken,
        refreshToken,
        expires
      })
    } else {
      const { accessToken, refreshToken, expires } = await authenticationService.authenticate({
        email
      });

      res.send({
        accessToken,
        refreshToken,
        expires
      })
    }
  });
};
```

Front the front-end I'm making a call like this:

```javascript
const manage = await axios.post(`${API_URL}/oauth2/azure`, {
    email: token?.email,
    firstName: token?.firstName,
    lastName: token?.lastName,
  }, {
    headers: {
      'authorization': `Bearer ${res?.data?.data?.auth_login?.access_token}`
    }
  })

accessToken = manage?.data?.accessToken
```

**Reply by NilsBaumgartner1994** [reply]
How did you get this done ? ```Once you get the token from the front-end OAuth```

**Reply by nodeworks** [reply]
i’m using next auth with next js. that assures the user is who they say they are, take the email from the response and do all the POST logic i described in the comment.

**Reply by luisfontes** [reply]
@nodeworks are you still using this method and does it work in rc.101?

I'm trying to make a similar flow, oAuth in next-auth and a custom endpoint in Directus for login/signup, but couldn't make it work. It seems the required method/params for the authenticate service changed.

**Reply by nodeworks** [reply]
Yup! Here's a crude example in rc.101. However, I did have to make some changes. I now pass in a token to insert into the session table. Basically it's a random uuid I pass from the front-end after next-auth.

```javascript
module.exports = function registerEndpoint(router, { services, exceptions, database, getSchema }) {
  router.post('/azure', async (req, res, next) => {
    const { ForbiddenException } = exceptions;
    const { UsersService } = services;

    if (!req?.token) {
      return next(new ForbiddenException());
    }

    const schema = await getSchema();
    if (!req?.body?.email) {
      res.send(false)
      return
    }

    const email = req?.body?.email.toLowerCase()
    const expiresDateTime = new Date(Date.now() + 7200000)
    const accessToken = req?.body?.token

    const user = await database
      .select('id', 'status')
      .from('directus_users')
      .whereRaw('LOWER(??) = ?', ['email', email])
      .first();

    const accountability = {
      ip: req.ip,
      userAgent: req.get('user-agent'),
      role: null,
    };

    // const mailService = new MailService({
    //   accountability,
    //   schema: req.schema,
    //   knex: database
    // });
    //
    // await mailService.send({
    //   to: email,
    //   subject: subjectLine,
    //   template: {
    //     name: 'user-invitation',
    //     data: {
    //       url: inviteURL.toString(),
    //       email,
    //     },
    //   },
    // });

    if (!user) {
      const usersService = new UsersService({
        schema,
        knex: database
      });

      await usersService.createOne({
        email,
        role: '64CFF9A9-9A5F-48F1-80D1-FF44F4CED284',
        first_name: req?.body?.firstName,
        last_name: req?.body?.lastName
      })
    }

    await database('directus_sessions').insert({
      token: accessToken,
      user: user.id,
      expires: expiresDateTime,
      ip: accountability?.ip,
      user_agent: accountability?.userAgent
    });

    res.send({
      userId: user.id,
      secret: process?.env?.SECRET
    })
  });
};

```

**Reply by nodeworks** [reply]
You may also have to add "external_identifier" when you create a new user. I forgot to add that field.

**Reply by fahami** [reply]
Hi @nodeworks, currently I am using the latest directus and can't find `authenticationService.authenticate`, may I know what method that can I use to implement similar behavior?

**Comment by liho00** [comment]
I think this feature quite helpful to handle authentication...

**Comment by NilsBaumgartner1994** [comment]
I propose either an implementation in the directus sdk directly or a new npm package.


Beste Grüße,
Nils Baumgartner

**Comment by rafazafar** [comment]
Was bit suprised to see this isnt included for a BaaS. Defeinitely a cleaner SSO implementation in SDK will help adoption. Currently using workaround via firebase as im too lazy to manually handle multiple SSOs.

**Reply by benhaynes** [reply]
Nothing from the team on this one yet... thoughts, @directus/team?

**Reply by rijkvanzanten** [reply]
It's really opinionated on how that could/would potentially work. In this case, the SDK would have to redirect to and handle the response from the provider, which would mean it would only reliably work in a browser environment, which breaks the current universal nature of the SDK. There's also no general consensus on what the methods should look like, so a blanket "lets just add OAuth2 support to the SDK" isn't enough for any real discussion here 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by delivey** [comment]
Nothing new? Would be very helpful.

**Reply by BlackDahlia313** [reply]
https://docs.directus.io/self-hosted/sso.html#seamless-sso

Under here it says its possible via SDK

Heres an example of a composable using SDK to do it: https://github.com/becem-gharbi/nuxt-directus/blob/main/src/runtime/composables/useDirectusAuth.ts

**Reply by customautosys** [reply]
> https://docs.directus.io/self-hosted/sso.html#seamless-sso
> 
> Under here it says its possible via SDK
> 
> Heres an example of a composable using SDK to do it: https://github.com/becem-gharbi/nuxt-directus/blob/main/src/runtime/composables/useDirectusAuth.ts

But the loginWithProvider there redirects to the provider URL, and the login function is still using email and password. How do I authenticate to the JS SDK if I have already redirected back to my app and I have an ID token and access token from auth0 already, without using email and password (because the person may be logging in with something else like X or LinkedIn)?

I've tried to ask in Discord also, nobody is helping me.

**Comment by NilsBaumgartner1994** [comment]
You need to redirect the user to Directus custom endpoint. The custom endpoint takes the access token and redirects you to your application (checks if redirect is allowed) and adds the access token to the url parameter.Currently I Think the SDK does not support it by default. It works with Cookies, but on mobile you will need I think the Access Token for various tasks.Beste Grüße,Nils BaumgartnerAm 17.07.2024 um 07:22 schrieb Custom Automated Systems Pte Ltd ***@***.***>:﻿

https://docs.directus.io/self-hosted/sso.html#seamless-sso
Under here it says its possible via SDK
Heres an example of a composable using SDK to do it: https://github.com/becem-gharbi/nuxt-directus/blob/main/src/runtime/composables/useDirectusAuth.ts

But the loginWithProvider there redirects to the provider URL, and the login function is still using email and password. How do I authenticate to the JS SDK if I have already redirected back to my app and I have an ID token and access token from auth0 already, without using email and password (because the person may be logging in with something else like X or LinkedIn)?
I've tried to ask in Discord also, nobody is helping me.

**Reply by customautosys** [reply]
> You need to redirect the user to Directus custom endpoint. The custom endpoint takes the access token and redirects you to your application (checks if redirect is allowed) and adds the access token to the url parameter.Currently I Think the SDK does not support it by default. It works with Cookies, but on mobile you will need I think the Access Token for various tasks.Beste Grüße,Nils BaumgartnerAm 17.07.2024 um 07:22 schrieb Custom Automated Systems Pte Ltd ***@***.***>:﻿
> 
> https://docs.directus.io/self-hosted/sso.html#seamless-sso
> Under here it says its possible via SDK
> Heres an example of a composable using SDK to do it: https://github.com/becem-gharbi/nuxt-directus/blob/main/src/runtime/composables/useDirectusAuth.ts
> 
> But the loginWithProvider there redirects to the provider URL, and the login function is still using email and password. How do I authenticate to the JS SDK if I have already redirected back to my app and I have an ID token and access token from auth0 already, without using email and password (because the person may be logging in with something else like X or LinkedIn)?
> I've tried to ask in Discord also, nobody is helping me.
> 
> —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

Thanks a lot for pointing the way ahead, do you know of any example (particularly which works with auth0-vue)?

Is there any way to login using just an API without another redirect, or to just ajax call the custom endpoint using the id token / access token? That's because on Android Capacitor it doesn't actually redirect to another URL but the redirect is a custom scheme that calls a handler provided by the auth0-vue library.

**Comment by NilsBaumgartner1994** [comment]
Direct copy of our redirect script.

```
import {defineEndpoint} from ***@***.***/extensions-sdk';
import ms from 'ms';
import {DatabaseInitializedCheck} from "../helpers/DatabaseInitializedCheck";
import {AppSettingsService} from "../helpers/ItemsServiceCreator";
import {StringHelper} from "../helpers/StringHelper";

const SCHEDULE_NAME = "redirect_with_token";
const env = process.env;
const PUBLIC_URL = env.PUBLIC_URL; // e.g. http://rocket-meals.de/rocket-meals/api or empty string

const WILDCARD = "\\*"; // we need to escape the * as it is a special character in regex
const WILDCARD_REPLACEMENT = "WILDCARD_REPLACEMENT";

function startsWithUntilWildcardReplacement(inputStr: string, whitelistStrWithWildcardReplacement: string): boolean {
    // redirectUrl could be "/rocket-meals/login/.../"
    // redirect_whitelist_entry_with_possible_wildcard_replacement could be "/rocket-meals/WILDCARD_REPLACEMENT/.../"
    // we need to check if the redirectUrl starts with the redirect_whitelist_entry_with_possible_wildcard_replacement
    let split = whitelistStrWithWildcardReplacement.split(WILDCARD_REPLACEMENT);
    let first_part: string = split.length > 1 ? split[0] as string : whitelistStrWithWildcardReplacement;
    // so /rocket-meals/ is the first part and the rest is the second part
    // remove the last char if it is a slash
    if(first_part.endsWith("/")){
       first_part = first_part.slice(0, -1);
    }
    // check if the redirectUrl starts with the first part
    return inputStr.startsWith(first_part);
}

// function to check if a url is allowed/matches a whitelist entry
function isRedirectUrlAllowedForWhitelistEntry(redirect_whitelist_entry: string, redirectUrl: URL): boolean {
    //console.log("isRedirectUrlAllowedForWhitelistEntry")
    //console.log("redirect_whitelist_entry: " + redirect_whitelist_entry)
    //console.log("redirectUrl: " + redirectUrl)
    // redirect_whitelist_entry is * then return true
    if (redirect_whitelist_entry === "*") {
       return true;
    }
    // redirectUrl could be
    // 127.0.0.1
    // 127.0.0.1/login?directus_refresh_token=1234
    // localhost
    // localhost:8081
    // localhost:8081/login?directus_refresh_token=1234
    // myapp://login?directus_refresh_token=1234
    // http://localhost:8081/login?directus_refresh_token=1234
    // https://myapp.com/login?directus_refresh_token=1234
    // myapp://*
    // http://localhost:8081/*
    // https://myapp.com/*

    // check if the redirectUrl matches the whitelist entry
    const redirectUrlProtocol = redirectUrl.protocol;
    const redirectUrlHost = redirectUrl.host;
    const redirectUrlPathname = redirectUrl.pathname;
    const redirectUrlSearch = redirectUrl.search;

    // as the redirect_whitelist_entry is a string, we need to check if it is a valid URL otherwise like "localhost" or "127.0.0.1" we need first to convert it to a valid URL
    // if no protocol is given, we assume http and https
    const hasProtocol = redirect_whitelist_entry.includes("://");
    if(!hasProtocol){
       //console.log("No protocol found in redirect_whitelist_entry: " + redirect_whitelist_entry)
       // if no protocol is given, we assume http and https
       let redirect_whitelist_entry_http = "http://" + redirect_whitelist_entry;
       let redirect_whitelist_entry_https = "https://" + redirect_whitelist_entry;
       // then we recursively call this function with the new URLs
       let redirect_allowed_http = isRedirectUrlAllowedForWhitelistEntry(redirect_whitelist_entry_http, redirectUrl);
       let redirect_allowed_https = isRedirectUrlAllowedForWhitelistEntry(redirect_whitelist_entry_https, redirectUrl);
       return redirect_allowed_http || redirect_allowed_https;
    } else {
       // okay we have a protocol which could be http, https, myapp, or a wildcard like *
       // we need to check if the protocol matches
       // if the redirectURL contains our WILDCARD_REPLACEMENT, then we reject it
       if(redirectUrlProtocol.includes(WILDCARD_REPLACEMENT)){
          //console.log("redirectUrlProtocol contains WILDCARD_REPLACEMENT, this is not allowed")
          return false;
       } else {
          //console.log("redirectUrl seems to be valid")
          // replace all * with WILDCARD_REPLACEMENT
          //console.log("Replace whitelist entry all wildcards with WILDCARD_REPLACEMENT")
          let replacedRedirect_whitelist_entry = StringHelper.replaceAll(redirect_whitelist_entry, WILDCARD, WILDCARD_REPLACEMENT);
          // create a new URL from the replaced string
          //console.log("replacedRedirect_whitelist_entry: " + replacedRedirect_whitelist_entry)
          let replacedRedirect_whitelist_entry_URL = new URL(replacedRedirect_whitelist_entry);
          //console.log("replacedRedirect_whitelist_entry_URL: " + replacedRedirect_whitelist_entry_URL)

          const replacedRedirect_whitelist_entry_URLProtocol = replacedRedirect_whitelist_entry_URL.protocol;
          const replacedRedirect_whitelist_entry_URLHost = replacedRedirect_whitelist_entry_URL.host;
          const replacedRedirect_whitelist_entry_URLPathname = replacedRedirect_whitelist_entry_URL.pathname;
          const replacedRedirect_whitelist_entry_URLSearch = replacedRedirect_whitelist_entry_URL.search;

          // check if the protocol matches
          const protocolMatches = startsWithUntilWildcardReplacement(redirectUrlProtocol, replacedRedirect_whitelist_entry_URLProtocol);
          if(!protocolMatches){
             return false;
          }

          // check if the host matches
          const hostMatches = startsWithUntilWildcardReplacement(redirectUrlHost, replacedRedirect_whitelist_entry_URLHost);
          if(!hostMatches){
             return false;
          }

          // check if the pathname matches
          const pathnameMatches = startsWithUntilWildcardReplacement(redirectUrlPathname, replacedRedirect_whitelist_entry_URLPathname);
          if(!pathnameMatches){
             return false;
          }

          // check if the search matches
          const searchMatches = startsWithUntilWildcardReplacement(redirectUrlSearch, replacedRedirect_whitelist_entry_URLSearch);
          if(!searchMatches){
             return false;
          }

          // if all checks passed, then the redirect URL is allowed
          return true;

       }
    }


}

function getValidUrl(url: string): URL | null {
    try {
       return new URL(url);
    } catch (e) {
       return null;
    }
}


// TO Test this Endpoint:
// 1. Login with a user in the Directus Admin UI
// 2. Go to the URL: http://127.0.0.1/rocket-meals/api/redirect-with-token?redirect=http://localhost:8081/login?directus_refresh_token=
// Where http://127.0.0.1/rocket-meals/api is the URL of the Directus API
export default defineEndpoint({
    id: 'redirect-with-token', // this will be the URL at which this endpoint is accessible
    handler: (router, {
       services,
       database,
       env,
       getSchema,
       logger
    }) => {


       router.get('/', async (req, res) => {
          let allTablesExist = await DatabaseInitializedCheck.checkAllTablesExist(SCHEDULE_NAME,getSchema);
          if (!allTablesExist) {
             return;
          }
          //console.log("#################################")
          //console.log("Redirect with token endpoint: settings")
          let redirectUrlIsValid = true;

          let redirect = req.query.redirect;

             if(!!redirect && typeof redirect === "string"){

                let redirectUrl = getValidUrl(redirect);

                if(!!redirectUrl) {
                   //let settings = await database(TABLENAME_FLOWHOOKS).first();
                   let schema = await getSchema();
                   let appSettingsService = new AppSettingsService(services, database, schema);
                   let settings = await appSettingsService.getAppSettings();
                   //console.log("App Settings")
                   //console.log(JSON.stringify(settings, null, 2))
                   let redirect_whitelist = settings?.redirect_whitelist;
                   if (!!redirect_whitelist) {
                      let foundValidRedirect = false;
                      if (redirect_whitelist.length === 0) {
                         foundValidRedirect = true; // no whitelist means all redirects are allowed
                      }

                      for (let i = 0; i < redirect_whitelist.length && !foundValidRedirect; i++) { // iterate over the whitelist as long as we haven't found a valid redirect
                         let redirect_whitelist_entry = redirect_whitelist[i];
                         try{
                            if (isRedirectUrlAllowedForWhitelistEntry(redirect_whitelist_entry, redirectUrl)) {
                               foundValidRedirect = true;
                               break;
                            }
                         } catch (e){
                            console.log("Error in redirect with token endpoint")
                            console.log("redirectUrl: " + redirectUrl)
                            console.log("redirect_whitelist_entry: " + redirect_whitelist_entry)
                            console.log(e)
                         }

                      }

                      redirectUrlIsValid = foundValidRedirect;
                   }
                } else {
                   redirectUrlIsValid = false; // no valid redirect URL found
                }
             } else {
                redirectUrlIsValid = false; // no redirect URL found
             }



          if(!redirectUrlIsValid){
             res.status(400).send("Invalid redirect URL");
             return;
          }

          //console.log("Cookies")
          //console.log(JSON.stringify(req.cookies, null, 2))

          const directus_refresh_token = req.cookies.directus_refresh_token;

          if(!!directus_refresh_token){ // this means the auth provider is using "cookie" mode.
             //console.log("directus_refresh_token found");

             const redirectURL = redirect + directus_refresh_token;
             //console.log("Redirect with token endpoint: redirectURL: " + redirectURL)
             res.redirect(redirectURL);
             return;
          }

          const directus_session_token = req.cookies.directus_session_token;
          if(!!directus_session_token) { // this means the auth provider is using "session" mode.
             // we need to obtain the directus_refresh_token from the directus_session_token
             //console.log("Redirect with token endpoint: directus_session_token: " + directus_session_token)
             const accountability = req?.accountability;
             const userId = req?.accountability?.user;
             //console.log("Redirect with token endpoint: userId: " + userId)
             if(!userId){
                res.status(400).send("No user found");
                return;
             }

             const knex = database;

             /**
              * Start of copy: https://github.com/directus/directus/blob/main/api/src/services/authentication.ts Login
              */

             const { nanoid } = await import('nanoid');

             const refreshToken = nanoid(64);
             const msRefreshTokenTTL: number = ms(String(env['REFRESH_TOKEN_TTL'])) || 0;
             const refreshTokenExpiration = new Date(Date.now() + msRefreshTokenTTL);

             await knex('directus_sessions').insert({
                token: refreshToken,
                user: userId,
                expires: refreshTokenExpiration,
                ip: accountability?.ip,
                user_agent: accountability?.userAgent,
                origin: accountability?.origin,
             });

             await knex('directus_sessions').delete().where('expires', '<', new Date());

             // End of copy

             const redirectURL = redirect + refreshToken;
             //console.log("Redirect with token endpoint: redirectURL: " + redirectURL)
             res.redirect(redirectURL);
             return;
          }

          // no token found
          res.status(400).send("No token found");
       });
    }
});
```

**Reply by customautosys** [reply]
I get the concept. Basically the key thing is to write a row into directus_sessions with the refreshToken. I will give that a try with the auth0 token. Thanks!

**Comment by NilsBaumgartner1994** [comment]
Could you then post a working example here?Beste Grüße,Nils BaumgartnerAm 21.07.2024 um 09:36 schrieb Custom Automated Systems Pte Ltd ***@***.***>:﻿
I get the concept. Basically the key thing is to write a row into directus_sessions with the refreshToken. I will give that a try with the auth0 token. Thanks!

**Reply by customautosys** [reply]
Apologies, I decided to give up as there was simply no way to register a new user without an email and password.

I tried to use a fake email and fake password, but then it would still be too insecure as the fake passwords have to be something fixed and either I cannot change them or I have to put in extra logic to change them periodically.

So I'm just using the directus collections to track the logins manually (basically I use 2 tables, 1 for a user object and the other to store the auth0 token subs for which 1 user could have several) and I am verifying the auth0 token manually too using jsonwebtoken. I am just using directus SDK only from the backend now and won't use directus for authenticating users but just for the admin interface and to call from my own backend. This also suits me because I need to restrict access to specific items in a collection depending on the user and I realise that directus doesn't allow that either.

**Comment by NilsBaumgartner1994** [comment]
Can the Directus SDK just use a https://developer.okta.com/blog/2018/12/13/oauth-2-for-native-and-mobile-apps PKCE Flow implement?

ADDITION:
```
My idea would be: Directus SDK creates random (v) then hashes it ($). SDK Generates a login url to the directus instance for a specific provider (google) and includes ($) the hash.
directus.com/auth/login/google?$=HashedRandomSecret & redirect=myMobileScheme
```

EXISITING FLOW
```
Directus instance then saves the has ($) waits redirects the user as normal to the provider (google).
google.com/clientID=skuef&....

At the provider the user authenticates, is then redirected to directus, which checks it. 
```

ADDITION:
```
Then directus redirects the user to a desired endpoint with (alpha) a generated code.
redirect: myMobileScheme/?alpha=PublicCodeServerGenerated

The user then takes (alpha) from the server, the random hash (v) and requests via POST the access_token?
POST: directus.com
{
   v: "RANDOM SECRET",
   alpha: "PublicCodeServerGenerated"
}
```

this is a scenario which i think would work on mobile for example with expo (react native). 

OPTIONAL:
```
We could also let the user send a "clientID" to make it more secure?
```

@rijkvanzanten What do you think of this in order to support OAuth for Mobile in a secure way?


<img width="687" alt="Bildschirmfoto 2024-08-23 um 14 34 44" src="https://github.com/user-attachments/assets/d8ff2c70-28bc-45a9-a83e-eac52f66e91f">

**Comment by NilsBaumgartner1994** [comment]
I combined solutions and other discussions here as an issue: https://github.com/directus/directus/issues/23468

**Reply by NilsBaumgartner1994** [reply]
Have a look at our PKCE solution?: https://github.com/directus/directus/discussions/23611
